`make release` implementation

mkolesnik opened the pull request:
These commits implement `make release`

Resolves: #123

mkolesnik committed:
Rename `create-release` to `do-release`

To aviod any confusion with upcoming `make release`

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Initial skeleton for `make release`

Also added semantic version validation to `make validate`

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Implement intial step of creating the file

Initially we create the release file with eaither statu shipyard, or set
to branch out in case of GA.
Made release-notes optional since this command won't fill them

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Added advancement code to the `release` target

This allows the target to advance the release yaml each time
`make release` is called.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Added GH PR creation to `make release`

The process will now automatically create a pull request to GitHub with
the current stage data.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Update README with release process

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt commented on Makefile:
```diff
@@ -55,6 +63,13 @@ Makefile.dapper:
 include Makefile.dapper
 
+git-sync:
+	git fetch --all --tags || :
```
```suggestion
	-git fetch --all --tags
```

mkolesnik committed:
Allow specifying GitHub account

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt reviewed (commented):
Hmm, I had started reviewing but this didn’t make it somehow :-/. (Review ongoing.)

skitt commented on scripts/lib/utils:
```diff
@@ -69,3 +80,24 @@ function checkout_project_branch() {
 }
 
+function is_semver() {
+    local ver="(0|[1-9]\d*)"
```
This fails *e.g.* with “0.10.0-m2”, because `\d` doesn’t represent `[[:digit:]]`.

```suggestion
    local ver="(0|[1-9][0-9]*)"
```

skitt commented on scripts/lib/utils:
```diff
@@ -69,3 +80,24 @@ function checkout_project_branch() {
+    semver['patch']=${patch_pre%%-*}
+    if [[ ${patch_pre} =~ - ]]; then
+        semver['pre']=${patch_pre#*-}
+    fi
```
Assuming that the value has been previously checked by `is_semver` (to enforce the field separator order):

```suggestion
    IFS=".-" read -r semver['major'] semver['minor'] semver['patch'] semver['pre'] <<<"$1"
```

skitt reviewed (changes requested):
To make it easier to test, since a dry run doesn’t need `GITHUB_TOKEN`, I changed `validate` in `release.sh` to specify run `dryrun expect_env GITHUB_TOKEN`.

Everything works with the requested changes, the release goes through the various stages. Thanks!

skitt commented on scripts/release.sh:
```diff
@@ -2,288 +2,154 @@
-# shellcheck source=scripts/lib/utils
-. "${DAPPER_SOURCE}/scripts/lib/utils"
-. "${SCRIPTS_DIR}/lib/debug_functions"
+source ${DAPPER_SOURCE}/scripts/lib/image_defs
+source ${DAPPER_SOURCE}/scripts/lib/utils
+source ${SCRIPTS_DIR}/lib/debug_functions
```
Did commits get mixed up here?

mkolesnik replied:
If you go commit-by-commit you'll see the very first one renames `release.sh` to `do-release.sh` and that gets picked up by GitHub correctly:

![image](https://user-images.githubusercontent.com/51078027/123223688-c7340900-d4d9-11eb-9851-8b65aff8fac6.png)

Unfortunately, in the "unified PR view" it looks as if I'm rewriting the whole file, and makes it confusing.

I suggest to follow commit-by-commit which should be clearer.

skitt replied:
Yes, I reviewed the code locally which made it easier to understand. What misled me initially was the difference in style at the start of `do-release.sh` and `release.sh`!

mkolesnik committed:
Added testing for `make release`

Added some testing to make sure `make release` works.

Also fixed some bugs the tests uncovered.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt reviewed (commented):
Wouldn’t it be better to retro-actively fix the faulty commits?

mkolesnik commented:
> Wouldn’t it be better to retro-actively fix the faulty commits?

It's possible though that would necessitate a force push
